Elevel/Elevel.Infrastructure.Services/Implementation/MailService.cs

```diff
@@ -0,0 +1,203 @@
+            catch (Exception ex)
+            {
+                return ex.Message;
+            }
+            finally
+            {
+                Disconnect();
+            }
+            return "Email was sent successfully";
+        }
+
+        private string UsersEmailNotification(List<Guid> receiverIds, string subject, string body)
+        {
+            var receiverEmails = _userManager.Users.Where(x => receiverIds.Contains(x.Id)).Select(x => x.Email).ToList();
+            if (receiverEmails == null)
+            {
+                return $"There are no valid emails";
+            }
+
+            _message.From.Add(new MailboxAddress("Elevel Notification", _emailConfiguration.Email));
+            _message.Body = new TextPart("plain")
+            {
+                Text = body
+            };
+            _message.Subject = subject;
+
+
+            foreach (var receiverEmail in receiverEmails)
+            {
+                _message.To.Add(MailboxAddress.Parse(receiverEmail));
+
+                try
+                {
+                    _smtpClient.Send(_message);
+                }
+                catch (Exception ex)
+                {
+                    Disconnect();
+                    return ex.Message;
+                }
+            }
+
+            return "Email was sent successfully";
+        }
+
+        private string MissedDeadlineEmailNotification(List<Test> tests, string subject, string body)
+        {
+            var UsersAndHrs = tests.GroupBy(x => x.HrId, (key, value) => new
+            {
+                HrId = key,
+                UserIds = value.Select(x => x.UserId)
+            });
+
+            _message.From.Add(new MailboxAddress("Elevel Notification", _emailConfiguration.Email));
+
+            _message.Subject = subject;
+
+
+            foreach (var pair in UsersAndHrs)
+            {
+                var hrEmail = _userManager.Users.FirstOrDefault(x => x.Id == pair.HrId).Email;
+                var Users = _userManager.Users.Where(x => pair.UserIds.Contains(x.Id));
+
+                if (hrEmail == null)
+                {
+                    return "Email was not sent";
+                }
+                if (Users == null)
+                {
+                    return "Email was not sent";
+                }
+
+                _message.To.Add(MailboxAddress.Parse(hrEmail));
+                foreach (var user in Users)
+                {
+                    _message.Body = new TextPart("plain")
+                    {
+                        Text = body
+                        .Replace("{FirstName}", user.FirstName, StringComparison.CurrentCultureIgnoreCase)
+                        .Replace("{LastName}", user.LastName, StringComparison.CurrentCultureIgnoreCase)
+                        .Replace("{Email}", user.Email, StringComparison.CurrentCultureIgnoreCase)
+                    };
+                    try
+                    {
+                        _smtpClient.Send(_message);
```

Comment:
Send in the loop it's not a good idea. Need revise this method. It can be dividing for few smaller methods.